### DISCUSSION TITLE
[Proposal] Allow parameter names with Func<> and Action<>

### DISCUSSION THREAD

**Discussion by SoftCircuits** [dotnet/csharplang#6262] [Language Ideas] [answered]
Consider a class with the following member.

    /// <summary>
    /// Returns part of the current string.
    /// </summary>
    public Func<int, int, string> Substring;

Such an approach might be taken if `Substring` requires different behavior depending on the current state of the parent object. So, this member could be set to reference different methods depending on that state.

The member is made public so that it can be accessed externally. This seems to work fine.

However, I have no way to indicate the purpose of the two `int` parameters. It seems like it would be helpful if I could do the following:

    /// <summary>
    /// Returns part of the current string.
    /// </summary>
    /// <param name="offset">The position of the first character to return.</param>
    /// <param name="length">Specifies the number of characters to return.</param>
    public Func<int offset, int length, string> Substring;

**Comment by HaloFour** [comment, marked as answer]
See:
#1883
#3703

**Comment by svick** [comment]
In this specific case, I think code like the following would be a better approach:

```c#
private Func<int, int, string> substring;

/// <summary>
/// Returns part of the current string.
/// </summary>
/// <param name="offset">The position of the first character to return.</param>
/// <param name="length">Specifies the number of characters to return.</param>
public string Substring(int offset, int length) => substring(offset, length);
```

It uses a more familiar API shape, can easily express the parameter names today and hides implementation details. I think all of that is worth a small increase in verbosity.

**Reply by SoftCircuits** [reply]
@svick Verbosity aside, your version is slightly less performant.

**Reply by svick** [reply]
It shouldn't. [JIT tends to be quite good at inlining trivial methods.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKBrIAIB1ASwwAsARGAGxgHNsGGDQDeNehPrEAzFICsAHmYA7DGnoq1U8gAYAfPQDKAV2DFdAbnGTrE8zvoBZABQBKW/THVJPqQHYjU3tncjRSVytvSQBfGljaagYWdkcYdggAE1EPAAcoZgA3QRh5JVV1TXV7A1wgyxoPGW0HEzNdZ016CAAzbtw0itV6XmU+dld6AF4aup1nHr6B4ZhR8YsfBqi7XSc3Dy9fSWIA1uDQ8MifePiaIA=) (Notice that the disassembly of both `M` methods is exactly the same.)

**Reply by SoftCircuits** [reply]
Wow, I never knew C# would inline code. I have a background in C++ and this is something that bothered me about C#.

In this case, performance was my main concern so that might not be an issue.

Thanks! I learned something new today. I need research this further.

**Reply by Xyncgas** [reply]
The people who are working on .NET and c# definitely uses c++ before to write their tools / runtime, although now they are saying switching to c# for everything, even if there is a question of chicken and egg, aside from that, I am generally in good faith that if I think something can be optimized they probably did it, but I always question if they are that smart, aside from that, having names about the type param in the meta data would be nice **however what if people want to use their own names**
